extensions/debug-server-ready/src/extension.ts

```diff
@@ -183,13 +191,104 @@ class ServerReadyDetector extends vscode.Disposable {
 		}
 	}
 
-	private debugWithBrowser(type: string, session: vscode.DebugSession, uri: string) {
+	private async debugWithBrowser(type: string, session: vscode.DebugSession, uri: string) {
+		const args = session.configuration.serverReadyAction as ServerReadyAction;
+		if (!args.killOnServerStop) {
+			await this.startBrowserDebugSession(type, session, uri);
+			return;
+		}
+
+		const trackerId = randomUUID();
+		const cts = new vscode.CancellationTokenSource();
+		const newSessionPromise = this.catchStartedDebugSession(session => session.configuration.trackerId === trackerId, cts.token);
+
+		if (!await this.startBrowserDebugSession(type, session, uri, trackerId)) {
+			cts.cancel();
+			cts.dispose();
+			return;
+		}
+
+		const createdSession = await newSessionPromise;
+		cts.dispose();
+
+		if (!createdSession) {
+			return;
+		}
+
+		const stopListener = vscode.debug.onDidTerminateDebugSession(async (terminated) => {
+			if (terminated === session) {
+				stopListener.dispose();
+				this.lateDisposables.delete(stopListener);
+				await vscode.debug.stopDebugging(createdSession);
+			}
+		});
+		this.lateDisposables.add(stopListener);
+	}
+
+	private startBrowserDebugSession(type: string, session: vscode.DebugSession, uri: string, trackerId?: string) {
 		return vscode.debug.startDebugging(session.workspaceFolder, {
 			type,
 			name: 'Browser Debug',
 			request: 'launch',
 			url: uri,
-			webRoot: session.configuration.serverReadyAction.webRoot || WEB_ROOT
+			webRoot: session.configuration.serverReadyAction.webRoot || WEB_ROOT,
+			trackerId,
```

Comment:
Please use a more unique name for this. Since it's being added to another extension's launch config, I think it should reflect where it came from and the fact that it's sort of internal, I suggest `_debugServerReadySessionId`.

Actually it's interesting that startDebugging doesn't return a DebugSession, I sort of expected it would. That would make this easier.

Reply:
It would indeed. But getting an API change in for that would be awkward, we'd need to duplicate the method 😕

Reply:
Done.